src/Tactic/Premises.agda

```diff
@@ -0,0 +1,294 @@
+               $ bundleHypotheses dhs′    -- (6)
+  let premiseTy  = tyView (is , quote ∃Dec ∙)
+      premiseTel = absTelescope is
+      premisePs  = map (λ (i , x) → arg (argInfo $ unAbs x) (` toℕ i))
+                       (uncurry L.zip $ map₁ reverse $ unzip $ enumerate is)
+      ∃premise?  = quote _,_ ◆⟦ premise ∣ quote it ∙ ⟧
+      cl         = clause premiseTel premisePs ∃premise?
+  print $ "```\n" <> show f <+> ":" <+> show premiseTy <> "\n"
+       -- <> show f <+> "⋯ =" <+> show premise <> " , it\n```\n"
+       -- TOO EXPENSIVE FOR LARGE RULES (even without `-v` flag)
+  declareDef (vArg f) premiseTy
+  defineFun f [ cl ]
+  where
+  -- (1) unfold syntactic sugar from Interface.STS
+  reduceRuleSyntax : Type → TC Type
+  reduceRuleSyntax ty = do
+    let whitelist = Interface.STS._∙_
+                 `∷ Interface.STS.∙_
+                 `∷ Interface.STS._────────────────────────────────_
+                 `∷ Interface.STS.────────────────────────────────_
+                 `∷ Interface.STS._───────────────────────────────────────_
+                 `∷ Interface.STS.───────────────────────────────────────_
+                 -- ^ include all syntactic sugar for inference rules here
+                 `∷ []
+    onlyReduceDefs whitelist $ go GAS_LIMIT ty
+    where mutual
+      go : ℕ → Type → TC Type
+      go 0 t = noGas t
+      go (suc n) = reduceDef >=> λ where
+        (Π[ x ∶ t ] ty) → do
+          t′  ← goArg n t
+          ty′ ← extendContext x t′ $ go n ty
+          return $ Π[ x ∶ t′ ] ty′
+        (def f as) → def f <$> goArgs n as
+        ty         → return ty
+
+      goArg : ℕ → Arg Term → TC (Arg Term)
+      goArg n (arg v t) = arg v <$> go n t
+
+      goArgs : ℕ → Args Term → TC (Args Term)
+      goArgs n = λ where
+        []       → return []
+        (a ∷ as) → _∷_ <$> goArg n a <*> goArgs n as
+
+      -- Gas-powered execution for non-terminating meta-programs.
+      GAS_LIMIT = 10
+
+      noGas : ∀ {A : Set} → A → TC A
+      noGas t = do
+        print $ "execution gas depleted" <+> "(GAS_LIMIT:" <+> show GAS_LIMIT <> ")"
+        return t
+
+  -- (2) separate implicit arguments from visible hypotheses
+  breakImplicits : AbsTelescope → AbsTelescope × List Type
+  breakImplicits = map₂  (unArgs ∘ map unAbs)
+                 ∘ break (isVisible? ∘ unAbs)
+
+  -- (3) unbundle hypotheses (always a single product due to `∙`)
+  --     * TODO: generalise to curried rules (which use `→`)
+  unbundleHypotheses : List Type → List Type
+  unbundleHypotheses = concatMap go where go = λ where
+    (def (quote _×_) (hArg _ ∷ hArg _ ∷ vArg x ∷ vArg ys ∷ [])) →
+      x ∷ go ys
+    t → [ t ]
+
+  -- (4) select only the hypotheses that are decidable
+  isDecidable? : AbsTelescope → Type → TC Bool
+  isDecidable? tel ty = runSpeculative $ (_, false) <$>
+    catchTC (checkType (quote it ∙) (tyView (tel , quote Dec ∙⟦ ty ⟧)) >> return true)
+            (return false)
```

Comment:
If you use the `MonadTC` typeclass from `stdlib-meta` you have the functions `runAndReset` and `isSuccessful`, and then you can write `runAndReset $ isSuccessful $ checkType (quote it ∙) (tyView (tel , quote Dec ∙⟦ ty ⟧))`. This suggests defining a function `hasInstance = runAndReset ∘ isSuccessful ∘ checkType (quote it ∙)`

Reply:
I'm having trouble using this interface:
1. I import the MonadTC interface to use `isSuccessful` instead:
```agda
open import Interface.MonadTC.Instance using (isSuccessful)
```
But I immediately get an error:
```bash
No instance of type MonadError (List ErrorPart) (λ {v} v₁ → TC v₁)
was found in scope.
when checking that the expression isSuccessful has type
(x : TC Term) → TC Bool
```

2. I then import the MonadError interface in hopes of fixing this, but to no avail:
```agda
open import Interface.MonadError.Instance
```

What else should I import?

Reply:
Yeah, the library design is sub-par here. The minimal thing would be this:
```
open import Interface.MonadTC using (Monad-TC; MonadTC-TC; module MonadTC)
open import Interface.MonadError using (MonadError-TC)

instance
  _ = Monad-TC
  _ = MonadTC-TC
  _ = MonadError-TC

open MonadTC ⦃...⦄ using (isSuccessful; runAndReset)
```